test: Verify server exit code during graceful process shutdown.

andriipatsula opened the pull request:
Fixes: https://github.com/scylladb/scylladb/issues/15365

# How to test:
Modify [main.cc](https://github.com/scylladb/scylladb/blob/master/main.cc#L123)
```C++
class stop_signal {
   void signaled() {
      // Causes abnormal program termination
      abort();
      // ...
   }
};
```
Run a test `./test.py --mode dev topology_custom/test_replace_ignore_nodes` or any other that relies on `manager.server_stop_gracefully(server_id)`

## Clean run (without changes to `scylla_cluster.py`):
> topology_custom   dev   [ **PASS** ] topology_custom.test_replace_ignore_nodes.1

The log file `test.py.dev.log` shows no issues.

## Modify `test/pylib/scylla_cluster.py` (apply the PR):
> topology_custom   dev   [ **FAIL** ] topology_custom.test_replace_ignore_nodes.1
...
The following test(s) have failed: topology_custom/test_replace_ignore_nodes
...
Server ScyllaServer(3, 127.68.82.3, e3ccf1b5-a8eb-4cfe-a322-0384babd1f02) exited with non-zero exit code: -6

The log file `test.py.dev.log` shows:
>22:43:44.078 INFO> [dev/topology_custom.test_replace_ignore_nodes.1] gracefully stopped ScyllaServer(3, 127.68.82.3, e3ccf1b5-a8eb-4cfe-a322-0384babd1f02)
22:43:44.079 ERROR> [dev/topology_custom.test_replace_ignore_nodes.1] Exception when executing _cluster_server_stop_gracefully: Server ScyllaServer(3, 127.68.82.3, e3ccf1b5-a8eb-4cfe-a322-0384babd1f02) exited with non-zero exit code: -6
Traceback (most recent call last):
...

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15660

andriipatsula commented:
/cc @kostja

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15660

andriipatsula commented:
> The commit message is empty -- once we merge the PR, only the commit message will remain, because this is a single-commit PR (so merge commit won't be created).

Updated to:
```
test: Verify server exit code during graceful process shutdown.

Update the test framework to monitor the server's returncode and throw an exception in the event of an abnormal server shutdown.
```

andriipatsula commented:
Just in case: It seems, the issue #15365 is a duplicate of #13205

> is an example of Scylla segfaulting in the middle of a test, but the test framework doesn't warn us about this. The segfault causes problems later and eventually a test failure.

avikivity commented:
> > The commit message is empty -- once we merge the PR, only the commit message will remain, because this is a single-commit PR (so merge commit won't be created).
> 
> Updated to:
> 
> ```
> test: Verify server exit code during graceful process shutdown.
> 
> Update the test framework to monitor the server's returncode and throw an exception in the event of an abnormal server shutdown.
> ```

Good practice is to explain why, not just what.

kbr-scylla commented:
The commit message still doesn't contain `Fixes scylladb/scylla#15365`.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=15660

andriipatsula commented:
> The commit message still doesn't contain `Fixes scylladb/scylla#15365`.

* Updated the commit message with the reason and added `Fixes scylladb/scylla#15365`
* Ran tests for both release and dev modes with `--repeat 20`. Ran tests for the debug mode with `--repeat 1` several times (for some reason, it takes "ages" to run with the `--repeat 20` parameter)

kbr-scylla commented:
Thanks, triggered CI

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4417/
- Duration: 2 hr 18 min
- Builder: spider4.cloudius-systems.com
